### PR TITLE
ENH: special: boostify F distribution functions

### LINKED ISSUE #7931: scipy.special.tests.test_basic::TestCephes::test_fdtri fails on i686
Test failure at commit 1106f4489badad63c20ab03d8513c58ebc34c2db
on 32-bit linux travis-ci: https://travis-ci.org/MacPython/scipy-wheels/jobs/279494135
```
=================================== FAILURES ===================================
____________________________ TestCephes.test_fdtri _____________________________

self = <scipy.special.tests.test_basic.TestCephes object at 0xe5a4bd6c>

    def test_fdtri(self):
        assert_allclose(cephes.fdtri(1, 1, [0.499, 0.5, 0.501]),
>                       array([0.9937365, 1.0, 1.00630298]), rtol=1e-6)

self       = <scipy.special.tests.test_basic.TestCephes object at 0xe5a4bd6c>

/venv/local/lib/python2.7/site-packages/scipy/special/tests/test_basic.py:366: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/venv/local/lib/python2.7/site-packages/numpy/testing/utils.py:1183: in assert_allclose
    verbose=verbose, header=header)
/venv/local/lib/python2.7/site-packages/numpy/testing/utils.py:607: in assert_array_compare
    chk_same_position(x_isnan, y_isnan, hasval='nan')
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

x_id = array([False,  True, False], dtype=bool)
y_id = array([False, False, False], dtype=bool), hasval = 'nan'

    def chk_same_position(x_id, y_id, hasval='nan'):
        """Handling nan/inf: check that x and y have the nan/inf at the same
            locations."""
        try:
            assert_array_equal(x_id, y_id)
        except AssertionError:
            msg = build_err_msg([x, y],
                                err_msg + '\nx and y %s location mismatch:' \
                                % (hasval), verbose=verbose, header=header,
                                names=('x', 'y'))
>           raise AssertionError(msg)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-06, atol=0
E           
E           x and y nan location mismatch:
E            x: array([ 0.9937365 ,         nan,  1.00630298])
E            y: array([ 0.9937365 ,  1.        ,  1.00630298])

err_msg    = ''
hasval     = 'nan'
header     = 'Not equal to tolerance rtol=1e-06, atol=0'
msg        = '\nNot equal to tolerance rtol=1e-06, atol=0\n\nx and y nan location mismatch:\n x: array([ 0.9937365 ,         nan,  1.00630298])\n y: array([ 0.9937365 ,  1.        ,  1.00630298])'
verbose    = True
x          = array([ 0.9937365 ,         nan,  1.00630298])
x_id       = array([False,  True, False], dtype=bool)
y          = array([ 0.9937365 ,  1.        ,  1.00630298])
y_id       = array([False, False, False], dtype=bool)

/venv/local/lib/python2.7/site-packages/numpy/testing/utils.py:587: AssertionError
```
Apparently the tests added in gh-7919 don't pass on 32-bit linux.

### LINKED ISSUE #20835: BUG: special.fdtri: inaccurate results for extreme arguments (exposed via scipy.stats.f.ppf)
### Describe your issue.

I suspect that there may be a defect in scipy.stats.f.ppf. I discovered this problem around 2021, before I joined the project, but I am reporting that I have recently received indications that the problem may remain.

The combination of the optimization function and the cdf is not consistent with the answer to the ppf function.

### Reproducing Code Example

```python
import scipy.stats as sps
import scipy.optimize as spo

print(sps.f.ppf(0.99999995, 1, 50000, loc=0, scale=1))
print(spo.brentq(lambda x: sps.f.cdf(x, 1, 50000) - 0.99999995, 1, 1e10))
```

### Error message

```shell
No error message is output. Normal completion, but the answer is questionable.
```

### SciPy/NumPy/Python version and system information

```shell
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT=0 DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= ZEN MAX_THREADS=64
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.27
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT=0 DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= ZEN MAX_THREADS=64
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.27
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.12.0
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.10
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  pythran:
    include directory: ../../tmp/pip-build-env-mnl4e8vy/overlay/lib/python3.10/site-packages/pythran
    version: 0.15.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /opt/python/cp310-cp310/bin/python
  version: '3.10'
```

### PR DESCRIPTION
**Pull request by dschmitz89**
#### Reference issue
Closes #20835 
Closes #7931

#### What does this implement/fix?
Use boost as backend for `fdtr`, `fdtrc` and `fdtri`. As these mostly rely on incomplete beta computations, boost should be much more accurate. If required, I can run an extensive accuracy test.

#### Additional information
Using boost's F distribution directly was recommended by boost maintainer @jzmaddock.

Potential follow ups could be to use boost also for the pdf and inverse survival function.

### CODE REVIEW

**Comment by dschmitz89** [comment]
Argh, that submodule hickup did not stop the build locally, it was not noticeable at all actually. Will fix later.

File: scipy/special/boost_special_functions.h [hunk 1]
```diff
@@ -1767,4 +1767,147 @@ landau_isf_double(double p, double loc, double scale)
...
+	return 1.0;
+    }
+    Real y;
+    try {
+	y = boost::math::cdf(
+                boost::math::fisher_f_distribution<Real, SpecialPolicy>(dfn, dfd), x);
+    } catch (...) {
+	/* Boost was unable to produce a result. Can happen when one or both
+	 * of v1 and v2 is very small and x is very large. e.g.
+	 * ncdtr(1e-100, 3, 1.5, 1e100). */
```
Comment (dschmitz89) [inline scipy/special/boost_special_functions.h:1793]:
```suggestion
	/* Boost was unable to produce a result. */
```

File: scipy/special/boost_special_functions.h [hunk 2]
```diff
@@ -1767,4 +1767,147 @@ landau_isf_double(double p, double loc, double scale)
...
+        sf_error("fdtr", SF_ERROR_NO_RESULT, NULL);
+        y = NAN;
+    }
+    if ((y < 0) || (y > 1)) {
+	/* Boost can return results far out of bounds when dfd and dfn are both large
+	 * and of similar magnitude. Return NAN if the result is out of bounds because
+	 * the answer cannot be trusted. */
```
Comment (dschmitz89) [inline scipy/special/boost_special_functions.h:1800]:
```suggestion
```

File: scipy/special/boost_special_functions.h [hunk 3]
```diff
@@ -1767,4 +1767,147 @@ landau_isf_double(double p, double loc, double scale)
...
+	return 0.0;
+    }
+    Real y;
+    try {
+	y = boost::math::cdf(boost::math::complement(
+                boost::math::fisher_f_distribution<Real, SpecialPolicy>(dfn, dfd), x));
+    } catch (...) {
+	/* Boost was unable to produce a result. Can happen when one or both
+	 * of v1 and v2 is very small and x is very large. e.g.
+	 * ncdtr(1e-100, 3, 1.5, 1e100). */
```
Comment (dschmitz89) [inline scipy/special/boost_special_functions.h:1888]:
```suggestion
```

File: scipy/special/boost_special_functions.h [hunk 4]
```diff
@@ -1767,4 +1767,147 @@ landau_isf_double(double p, double loc, double scale)
...
+        sf_error("fdtrc", SF_ERROR_NO_RESULT, NULL);
+        y = NAN;
+    }
+    if ((y < 0) || (y > 1)) {
+	/* Boost can return results far out of bounds when dfd and dfn are both large
+	 * and of similar magnitude. Return NAN if the result is out of bounds because
+	 * the answer cannot be trusted. */
```
Comment (dschmitz89) [inline scipy/special/boost_special_functions.h:1895]:
```suggestion
```

File: scipy/special/cython_special.pyx [hunk 5]
```diff
@@ -3134,6 +3112,33 @@ cpdef df_number_t ncfdtr(df_number_t x0, df_number_t x1, df_number_t x2, df_numb
...
+    """See the documentation for scipy.special.fdtr"""
+    if df_number_t is float:
+        return (<float(*)(float, float, float) noexcept nogil>scipy.special._ufuncs_cxx._export_f_cdf_float)(x0, x1, x2)
+    elif df_number_t is double:
+        return (<double(*)(double, double, double) noexcept nogil>scipy.special._ufuncs_cxx._export_f_cdf_double)(x0, x1, x2)
+    else:
+        return NAN
+
+cpdef df_number_t fdtrc(df_number_t x0, df_number_t x1, df_number_t x2) noexcept nogil:
+    """See the documentation for scipy.special.fdtr"""
```
Comment (dschmitz89) [inline scipy/special/cython_special.pyx:3125]:
```suggestion
    """See the documentation for scipy.special.fdtrc"""
```

File: scipy/special/cython_special.pyx [hunk 6]
```diff
@@ -3134,6 +3112,33 @@ cpdef df_number_t ncfdtr(df_number_t x0, df_number_t x1, df_number_t x2, df_numb
...
+    if df_number_t is float:
+        return (<float(*)(float, float, float) noexcept nogil>scipy.special._ufuncs_cxx._export_f_sf_float)(x0, x1, x2)
+    elif df_number_t is double:
+        return (<double(*)(double, double, double) noexcept nogil>scipy.special._ufuncs_cxx._export_f_sf_double)(x0, x1, x2)
+    else:
+        return NAN
+
+cpdef df_number_t fdtri(df_number_t x0, df_number_t x1, df_number_t x2) noexcept nogil:
+    """See the documentation for scipy.special.fdtr"""
```
Comment (dschmitz89) [inline scipy/special/cython_special.pyx:3134]:
```suggestion
    """See the documentation for scipy.special.fdtri"""
```

**Comment by dschmitz89** [comment]
CC @steppi : let me know if you would like to hold off on cephes to boost migrations until your `xsf` library is in place. I guess it is a nontrivial amount of work to port over changes from scipy to xsf back and forth. If `xsf` is still a bit further down the pipeline, this PR should be a low hanging fruit compared to many others in special.

**Comment by rgommers** [comment]
This is a large change and doesn't look ready for 1.16.0, ideally we'd not merge Boost-related changes right before a release branch anyway (corner cases and error handling policies tend to be fragile). So moving to 1.17.0

@steppi it'd be great if you answer the question from @dschmitz89 a few comments up.

**Comment by steppi** [comment]
> This is a large change and doesn't look ready for 1.16.0, ideally we'd not merge Boost-related changes right before a release branch anyway (corner cases and error handling policies tend to be fragile). So moving to 1.17.0
> 
> @steppi it'd be great if you answer the question from @dschmitz89 a few comments up.

Sorry for missing the question @dschmitz89 . This didn't need to wait for `xsf`. I just think it needs more thorough local testing (not necessarily more tests in the test suite). I just commented this here https://github.com/scipy/scipy/pull/22539#issuecomment-2888306011 too. Like @rgommers said, we've seen these migrations to Boost cause trouble due to corner cases and error handling policies pretty often.

**Comment by dschmitz89** [comment]
> > This is a large change and doesn't look ready for 1.16.0, ideally we'd not merge Boost-related changes right before a release branch anyway (corner cases and error handling policies tend to be fragile). So moving to 1.17.0
> > @steppi it'd be great if you answer the question from @dschmitz89 a few comments up.
> 
> Sorry for missing the question @dschmitz89 . This didn't need to wait for `xsf`. I just think it needs more thorough local testing (not necessarily more tests in the test suite). I just commented this here [#22539 (comment)](https://github.com/scipy/scipy/pull/22539#issuecomment-2888306011) too. Like @rgommers said, we've seen these migrations to Boost cause trouble due to corner cases and error handling policies pretty often.

Yup, will do. cephes is much better than cdflib, so let's see if we find cases where it outperforms boost. I do not expect that here though as boost's incomplete beta was superior to cephes by a large margin (#17697 ).

**Comment by dschmitz89** [comment]
The benchmarks show that this PR is a definite improvement, so I am going ahead with merging it. If issues arise in downstream testing, I will make time to fix them up.
